maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/checkout/GitSshCheckOutCommandTckTest.java

```diff
@@ -0,0 +1,185 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.file.FileSystem;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.security.GeneralSecurityException;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.security.SecureRandom;
+import java.util.Collections;
+
+import org.apache.maven.scm.provider.ScmProviderRepositoryWithHost;
+import org.apache.maven.scm.provider.git.GitScmTestUtils;
+import org.apache.maven.scm.repository.ScmRepository;
+import org.apache.maven.scm.tck.command.checkout.CheckOutCommandTckTest;
+import org.apache.sshd.common.config.keys.KeyUtils;
+import org.apache.sshd.git.GitLocationResolver;
+import org.apache.sshd.git.pack.GitPackCommandFactory;
+import org.apache.sshd.server.SshServer;
+import org.apache.sshd.server.auth.pubkey.KeySetPublickeyAuthenticator;
+import org.apache.sshd.server.auth.pubkey.PublickeyAuthenticator;
+import org.apache.sshd.server.session.ServerSession;
+import org.apache.sshd.util.test.CommonTestSupportUtils;
+import org.apache.sshd.util.test.CoreTestSupportUtils;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.openssl.PKCS8Generator;
+import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
+import org.bouncycastle.openssl.jcajce.JcaPKCS8Generator;
+import org.bouncycastle.openssl.jcajce.JceOpenSSLPKCS8EncryptorBuilder;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.OutputEncryptor;
+import org.bouncycastle.util.io.pem.PemObject;
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+/**
+ *
+ */
+public abstract class GitSshCheckOutCommandTckTest
+    extends CheckOutCommandTckTest
+{
+    protected final SshServer sshServer;
+    protected final KeyPair keyPair;
+
+    @Rule
+    public TemporaryFolder tmpDirectory = new TemporaryFolder();
+
+    protected GitSshCheckOutCommandTckTest() throws GeneralSecurityException
+    {
+         sshServer = CoreTestSupportUtils.setupTestServer( getClass() );
+         keyPair = CommonTestSupportUtils.generateKeyPair( KeyUtils.RSA_ALGORITHM, 2048 );
+         PublickeyAuthenticator authenticator = new KeySetPublickeyAuthenticator( "onlykey",
+                 Collections.singletonList( keyPair.getPublic() ) );
+         sshServer.setPublickeyAuthenticator( authenticator );
+    }
+
+    static void writePrivateKeyAsPkcs8( Path file, PrivateKey privateKey, String passphrase )
+            throws IOException, OperatorCreationException
+    {
+        final OutputEncryptor encryptor;
+        // encryption only optional
+        if ( passphrase != null )
+        {
+            //encrypted form of PKCS#8 file  
```

Comment:
@michael-o If you have any idea, why the private key file is not accepted by JGit and Gitexe we could fix it and enable also the test with passphrase.

Reply:
I cannot reproduce this. With `-Prun-its` it works on Windows 10 and FreeBSD 12 with Java 8 and 11. All tests pass...

Reply:
Weird:
```
sh://127.0.0.1:52277/repository working-copy"
[main] INFO org.apache.maven.scm.provider.git.gitexe.command.GitCommandLineUtils - Working directory: D:\Entwicklung\Projekte\maven-scm\maven-scm-providers\maven-scm-providers-git\maven-scm-provider-gitexe\target\scm-test
Exception in thread "Thread-682" java.lang.IllegalStateException: Executor has been shut down
        at org.apache.sshd.common.util.ValidateUtils.createFormattedException(ValidateUtils.java:213)
        at org.apache.sshd.common.util.ValidateUtils.throwIllegalStateException(ValidateUtils.java:207)
        at org.apache.sshd.common.util.ValidateUtils.checkState(ValidateUtils.java:184)
        at org.apache.sshd.common.util.threads.NoCloseExecutor.execute(NoCloseExecutor.java:100)
        at java.base/sun.nio.ch.AsynchronousChannelGroupImpl.executeOnPooledThread(AsynchronousChannelGroupImpl.java:188)
        at java.base/sun.nio.ch.Invoker.invokeIndirectly(Invoker.java:213)
        at java.base/sun.nio.ch.Invoker.invokeIndirectly(Invoker.java:314)
        at java.base/sun.nio.ch.WindowsAsynchronousServerSocketChannelImpl$AcceptTask.failed(WindowsAsynchronousServerSocketChannelImpl.java:286)
        at java.base/sun.nio.ch.Iocp$EventHandlerTask.run(Iocp.java:389)
        at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 5.221 s <<< FAILURE! - in org.apache.maven.scm.provider.git.gitexe.command.checkout.GitExeSshCheckOutCommandTckTest
[ERROR] testCheckOutCommandTestWithPassphrase(org.apache.maven.scm.provider.git.gitexe.command.checkout.GitExeSshCheckOutCommandTckTest)  Time elapsed: 1.944 s  <<< FAILURE!
java.lang.AssertionError:
Check result was successful, output: Cloning into 'working-copy'...
Warning: Permanently added '[127.0.0.1]:********@127.0.0.1: Permission denied (password,keyboard-interactive,publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.apache.maven.scm.ScmTckTestCase.checkOut(ScmTckTestCase.java:202)
```
Is the first exception expected?

Reply:
Stored key pairs often lead to false positives with security scanning (regularly performed by GitHub), therefore I would be reluctant to commit SSH keys.

Reply:
Ok, I changed my mind and added a test for an encrypted OpenSSH private key being created with `ssh-keygen` in https://github.com/apache/maven-scm/pull/155/commits/be6124ee0a73810ee622e24caed0bf4dcc522013 and it passes.

Still don't know why the dynamically created private key does not work, though.

Reply:
Dynamically created encrypted private key fixed now in https://github.com/apache/maven-scm/pull/155/commits/6d7b05b6223d39dd9aad0d1b225078037b8723ac.